File: 1458. Max Dot Product of Two Subsequences.py
class Solution:
    def maxDotProduct(self, nums1: list[int], nums2: list[int]) -> int:
        dp = [[0]*(len(nums2)) for _ in range(len(nums1))]

        a_, *nums1 = nums1
        b_, *nums2 = nums2

        dp[0][0] = a_ * b_
        
        for i, a in enumerate(nums1, 1):
            dp[i][0] = max(dp[i-1][0], a * b_)
        
        for j, b in enumerate(nums2, 1):
            dp[0][j] = max(dp[0][j-1], b * a_)

        for i, a in enumerate(nums1, 1):
            for j, b in enumerate(nums2, 1):
                prod = a * b
                dp[i][j] = max(prod, 
                               prod + dp[i-1][j-1],
                               dp[i-1][j], 
                               dp[i][j-1], 
                               )
        return dp[-1][-1]

File: 1458. Max Dot Product of Two Subsequences.java
class Solution {
    public int maxDotProduct(int[] nums1, int[] nums2) {
        if(nums2.length<nums1.length) return maxDotProduct(nums2,nums1);
        int l1=nums1.length,l2=nums2.length;
        int[] memo=new int[l1];
        for(int n: nums2) {
            int diag=0,last=0;
            for(int i=0;i<l1;i++) {
                int tmp=memo[i];
                last=memo[i]=Math.max(last,Math.max(tmp,diag+n*nums1[i]));
                diag=tmp;
            }
        }
        
        if(memo[l1-1]==0) {
            int max1=5000,max2=5000;
            for(int n: nums1) if(Math.abs(n)<Math.abs(max1)) max1=n;
            for(int n: nums2) if(Math.abs(n)<Math.abs(max2)) max2=n;
            return max1*max2;
        }
        return memo[l1-1];
    }
}
